### PR TITLE
[ADD][9.0] medical_prescription_merge: Rx order merging

### PR DESCRIPTION
**Pull request by obulkin**
* Add medical.prescription.order.merge wizard model with logic for merging any
number of existing prescription orders into one and various validations
related to this process. Users select a destination order whose info takes
precedence when there are conflicts and can choose to skip certain validations.
* Add a form view, a menu item, and an action for the wizard model

### CODE REVIEW

File: medical_prescription_merge/wizards/medical_prescription_order_merge.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        default=lambda s: s._compute_default_merge_orders(),
+    )
+    dest_order_id = fields.Many2one(
+        comodel_name='medical.prescription.order',
+        string='Destination Order',
+        help='This prescription order will serve as the baseline for the merge'
+             ' (its info will take precedence when there are conflicts)',
+        required=True,
+        default=lambda s: s._compute_default_dest_order(),
+        # domain='[("id", "in", merge_order_ids[0][2])]',
```
Comment (obulkin) [inline medical_prescription_merge/wizards/medical_prescription_order_merge.py]: @lasley - As per our convo earlier, the wizard should only allow users to select a destination order that is in the merge order list. The domain shown here will do that, but you decided that it was too hacky and wanted to look into it further as part of the PR review.

File: medical_prescription_merge/wizards/medical_prescription_order_merge.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    )
+    skip_validation = fields.Boolean(
+        help='Check this box to allow prescription orders issued by different'
+             ' physicians or for different dates to be merged',
+        string='Skip Validation?',
+    )
+
+    @api.model
+    def _compute_default_merge_orders(self):
+        if self.env.context.get('active_model') \
```
Comment (lasley) [inline medical_prescription_merge/wizards/medical_prescription_order_merge.py]: `\` is a nasty hack, use an intermediary var instead

```
        active_model = self.env.context.get('active_model')
        if active_model == 'medical.prescription.order':
```

File: medical_prescription_merge/wizards/medical_prescription_order_merge.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                == 'medical.prescription.order':
```
Comment (lasley) [inline medical_prescription_merge/wizards/medical_prescription_order_merge.py]: `rx_model = 'medical.prescription.order'` and remove all the duplicate strings

File: medical_prescription_merge/wizards/medical_prescription_order_merge.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            return self.env['medical.prescription.order'].browse(
+                self.env.context.get('active_ids')
```
Comment (lasley) [inline medical_prescription_merge/wizards/medical_prescription_order_merge.py]: Another place that intermediary var would help

File: medical_prescription_merge/wizards/medical_prescription_order_merge.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            )
+        else:
+            return self.env['medical.prescription.order']
+
+    @api.model
+    def _compute_default_dest_order(self):
+        merge_order_ids = self._compute_default_merge_orders()
+        if merge_order_ids:
```
Comment (lasley) [inline medical_prescription_merge/wizards/medical_prescription_order_merge.py]: Better to ask forgiveness than permission:

```
        try:
            return self._compute_default_merge_orders()[0]
        except (IndexError, TypeError):
            return
```

File: medical_prescription_merge/wizards/medical_prescription_order_merge.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    def action_merge(self):
+        self.ensure_one()
+
+        if len(self.merge_order_ids) < 2:
+            raise ValidationError(_(
+                'You must select at least two orders to start a merge.'
+            ))
+
+        if not self.skip_validation:
+            merge_physicians = set(self.merge_order_ids.mapped('physician_id'))
```
Comment (lasley) [inline medical_prescription_merge/wizards/medical_prescription_order_merge.py:79]: Move this into a method of its own `_do_validation` or similar. Allows for inherited modules to inject logic. Rule of thumb, logic in loops is inaccessible to downstream modules. This definitely should be accessible.

**Review by lasley — CHANGES_REQUESTED** [review]

File: medical_prescription_merge/wizards/medical_prescription_order_merge.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        domain=lambda s: s._compute_dest_order_domain(),
+    )
+    skip_validation = fields.Boolean(
+        help='Check this box to allow prescription orders issued by different'
+             ' physicians or for different dates to be merged',
+        string='Skip Validation?',
+    )
+
+    @api.model
+    def _compute_default_merge_orders(self):
```
Comment (lasley) [inline medical_prescription_merge/wizards/medical_prescription_order_merge.py]: _default_merge_orders

File: medical_prescription_merge/wizards/medical_prescription_order_merge.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        if self.env.context.get('active_model') \
+                == 'medical.prescription.order':
+            return self.env['medical.prescription.order'].browse(
+                self.env.context.get('active_ids')
+            )
+        else:
+            return self.env['medical.prescription.order']
+
+    @api.model
+    def _compute_default_dest_order(self):
```
Comment (lasley) [inline medical_prescription_merge/wizards/medical_prescription_order_merge.py]: _default_dest_order

**Review by lasley — CHANGES_REQUESTED** [review]

**Comment by lasley** [comment]
@obulkin all your's

**Comment by lasley** [comment]
Oops it seem my push failed. Now it's all your's

File: medical_prescription_merge/wizards/medical_prescription_order_merge.py (hunk 6 above)
Comment (obulkin) [reply medical_prescription_merge/wizards/medical_prescription_order_merge.py:79]: Sure thing. There are no loops happening here, but all the same, the code doesn't provide a way for child classes to overload the validations, and I definitely see the value in that.

Comment (lasley) [reply medical_prescription_merge/wizards/medical_prescription_order_merge.py:79]: Oops I meant to say control structure, not loop. Logic in a control structure is inaccessible. Sounds like you got that though 😄

**Review by lasley — APPROVED** [review]

**Pull request merged**